File: compiler/src/typecheck/bounds.rs
use crate::{
    database::{Db, Fact, NodeRef, Render},
    nodes::TraitDefinitionNode,
    typecheck::Substitutions,
};
use std::collections::BTreeMap;

#[derive(Debug, Clone)]
pub struct Instance {
    pub node: NodeRef,
    pub trait_node: NodeRef,
    pub substitutions: Substitutions,
    pub from_bound: bool,
    pub default: bool,
    pub error: bool,
}

#[derive(Debug, Clone, Default)]
pub struct Instances(pub Vec<Instance>);

impl Fact for Instances {}

impl Render for Instances {
    fn write(&self, w: &mut dyn std::fmt::Write, _db: &Db) -> std::fmt::Result {
        write!(w, "has instances")
    }
}

#[derive(Debug, Clone, Default)]
pub struct Bounds(pub BTreeMap<NodeRef, BoundsItem>);

impl Fact for Bounds {}

impl Render for Bounds {
    fn write(&self, w: &mut dyn std::fmt::Write, db: &Db) -> std::fmt::Result {
        if self.0.is_empty() {
            return write!(w, "has bound(s)");
        }

        write!(w, "has bound(s) ")?;

        for (index, (_, item)) in self.0.iter().enumerate() {
            if index > 0 {
                write!(w, ", ")?;
            }

            write!(w, "{}", db.render(&item.bound))?;

            write!(w, " (")?;
            if let Some(instance) = &item.instance {
                write!(w, "{}", db.render(&instance.instance_node))?;
            } else {
                write!(w, "unresolved")?;
            }
            write!(w, ")")?;
        }

        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct BoundsItem {
    pub bound: Bound,
    pub instance: Option<BoundsItemInstance>,
}

#[derive(Debug, Clone)]
pub struct BoundsItemInstance {
    pub instance_node: NodeRef,
    pub resolved_node: NodeRef,
    pub from_bound: bool,
    pub error: bool,
}

#[derive(Debug, Clone)]
pub struct Bound {
    pub source_node: NodeRef,
    pub bound_node: NodeRef,
    pub trait_node: NodeRef,
    pub substitutions: Substitutions,
    pub optional: bool,
}

impl Bound {
    pub fn to_instance(&self, node: NodeRef) -> Instance {
        Instance {
            node,
            trait_node: self.trait_node.clone(),
            substitutions: self.substitutions.clone(),
            from_bound: true,
            default: false,
            error: false,
        }
    }
}

impl Render for Bound {
    fn write(&self, f: &mut dyn std::fmt::Write, db: &Db) -> std::fmt::Result {
        let trait_definition = self
            .trait_node
            .downcast_ref::<TraitDefinitionNode>()
            .expect("not a trait definition");

        write!(f, "{}", trait_definition.name)?;

        for parameter in &trait_definition.parameters {
            let ty = self
                .substitutions
                .get(parameter)
                .expect("missing substitution");

            write!(f, " ")?;

            ty.render(false).write(f, db)?;
        }

        Ok(())
    }
}

File: compiler/src/typecheck/Bounds.swift
import OrderedCollections

public struct Instance {
    public var node: Node
    public var traitNode: Node
    public var substitutions: Substitutions
    public var isFromBound = false
    public var isDefault = false
    public var isError = false
}

public struct Instances: Fact {
    public var instances: [Instance] = []

    public func render(into context: RenderContext) { context.string("has instances") }
}

public struct Bounds: Fact {
    public var bounds: OrderedDictionary<Node, (Bound, ResolvedBound?)> = [:]

    public func render(into context: RenderContext) {
        guard !self.bounds.isEmpty else {
            context.string("has bound(s)")
            return
        }

        context.string("has bound(s) ")

        for (index, (_, (bound, instance))) in self.bounds.enumerated() {
            if index > 0 { context.string(", ") }

            bound.render(into: context)

            context.string(" (")
            if let resolved = instance {
                context.node(resolved.instance.node)
            } else {
                context.string("unresolved")
            }
            context.string(")")
        }
    }
}

public struct ResolvedBound {
    public let instance: Instance
    public let instanceSubstitutions: Substitutions
    public let resolvedNode: Node
    public let resolvedSubstitutions: Substitutions
}

public struct Bound {
    public var sourceNode: Node
    public var boundNode: Node
    public var traitNode: Node
    public var targetNode: Node?
    public var substitutions: Substitutions
    public var isOptional = false
}

extension Instance {
    public static func resolved(_ node: Node, from bound: Bound) -> Self {
        return .init(
            node: node,
            traitNode: bound.traitNode,
            substitutions: bound.substitutions,
            isFromBound: true,
        )
    }
}

extension Bound: Renderable {
    public func render(into context: RenderContext) {
        let traitDefinition =
            context.db[self.traitNode, Defined.self]!.definition as! TraitDefinition

        context.code(traitDefinition.name!)

        for parameter in traitDefinition.parameters {
            context.string(" ")
            if let type = self.substitutions[parameter] {
                type.render(into: context, root: false)
            } else {
                context.code("_")
            }
        }
    }
}
